### PR TITLE
[Master] fix15742

### LINKED ISSUE #15742: Generic JSX components
The fix in https://github.com/Microsoft/TypeScript/pull/15687 did not get the original issue resolved, now that we have the apparent type, we can not spread `T` into a `Component<T>`.

here are a few samples:

```ts
import * as React from "react";

const decorator = function <T>(Component: React.StatelessComponent<T>): React.StatelessComponent<T> {
    return (props) => <Component {...props} ></Component> // should be ok
};

const decorator2 = function <T extends { x: number }>(Component: React.StatelessComponent<T>): React.StatelessComponent<T> {
    return (props) => <Component {...props} x={2} ></Component>  // Should be ok, x is known
};

const decorator4 = function <T extends { x: number }>(Component: React.StatelessComponent<T>): React.StatelessComponent<T> {
    return (props) => <Component {...props} y={"blah"} ></Component>  // Should be an error, unknown y
};

```

### PR DESCRIPTION
**Pull request by yuit**
Fix #15742 
Note: There is still an issue when using generic with class see https://github.com/Microsoft/TypeScript/compare/master-fix15742?expand=1#diff-c5ae2fab8e9a760ddeda06b1b9128151

TODO

- [ ] port to release

- [x] add more tests

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -13548,6 +13548,17 @@ namespace ts {
...
 
+        function createIntersectionOfApparentTypeOfJsxPropsType(propsType: Type): Type {
+            if (propsType && propsType.flags & TypeFlags.Intersection) {
+                const propsApprentType: Type[] = [];
+                for (const t of (<UnionOrIntersectionType>propsType).types) {
+                    propsApprentType.push(getApparentType(t));
+                }
+                return getIntersectionType(propsApprentType);
+            }
+            return propsType;
```
Comment (mhegazy) [inline src/compiler/checker.ts]: what if it is not an intersection, you do not want to call getApparentType on it?

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -13548,6 +13548,20 @@ namespace ts {
             return _jsxElementChildrenPropertyName;
         }
 
+        function getApparentTypeOfJsxPropsType(propsType: Type): Type {
+            if (propsType) {
+                if (propsType.flags & TypeFlags.Intersection) {
+                    const propsApprentType: Type[] = [];
+                    for (const t of (<UnionOrIntersectionType>propsType).types) {
```
Comment (sandersn) [inline src/compiler/checker.ts]: doesn't `getApparentType` of an intersection type do this loop for you?

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -13548,6 +13548,20 @@ namespace ts {
             return _jsxElementChildrenPropertyName;
         }
 
+        function getApparentTypeOfJsxPropsType(propsType: Type): Type {
+            if (propsType) {
```
Comment (sandersn) [inline src/compiler/checker.ts]: I like this better: 
```ts
if (!propsType) {
  return undefined;
}
```

**Comment by sandersn** [comment]
Never mind, I looked more closely at `getApparentType` and it doesn't recursively call `getApparentType` on `types`.

File: src/compiler/checker.ts (hunk 2 above)
Comment (sandersn) [inline src/compiler/checker.ts]: Later: no it doesn't. :( You should be able to express this as `getIntersectionType(t.types.map(getApparentType)))` though, right?

**Comment by yuit** [comment]
@sandersn  I wasn't expected that as well

**Comment by mhegazy** [comment]
@yuit can you please port this to release-2.3

**Comment by mhegazy** [comment]
never mind, I ported it.

**Pull request merged**